python/tester/upload.py

```diff
@@ -84,34 +153,222 @@ def main():
         "./prompt-proto-upload.json"
     )
     storage_client = storage.Client(PROJECT_ID, credentials=credentials)
-    bucket = storage_client.bucket("rubin-prompt-proto-main")
+    dest_bucket = storage_client.bucket("rubin-prompt-proto-main")
     batch_settings = pubsub_v1.types.BatchSettings(
         max_messages=INSTRUMENTS[instrument].n_detectors,
     )
     publisher = pubsub_v1.PublisherClient(credentials=credentials,
                                           batch_settings=batch_settings)
 
+    last_group = get_last_group(storage_client, instrument, date)
+    _log.info(f"Last group {last_group}")
+
+    src_bucket = storage_client.bucket("rubin-prompt-proto-unobserved")
+    raw_pool = get_samples(src_bucket, instrument)
+
+    if raw_pool:
+        _log.info(f"Observing real raw files from {instrument}.")
+        for i, true_group in enumerate(itertools.islice(itertools.cycle(raw_pool), n_groups)):
+            group = last_group + i + 1
+            _log.debug(f"Processing group {group} from unobserved {true_group}...")
+            # snap_dict maps snap_id to {visit: blob}
+            snap_dict = {}
+            for snap_id, old_visits in raw_pool[true_group].items():
+                snap_dict[snap_id] = {splice_group(true_visit, group): blob
+                                      for true_visit, blob in old_visits.items()}
+            visit_infos = {info for det_dict in snap_dict.values() for info in det_dict}
+
+            # TODO: may be cleaner to use a functor object than to depend on
+            # closures for the bucket and data.
+            def upload_from_pool(visit, snap_id):
+                src_blob = snap_dict[snap_id][visit]
+                exposure_id = make_exposure_id(visit.instrument, visit.group, snap_id)
+                fname = raw_path(visit.instrument, visit.detector, visit.group, snap_id,
+                                 exposure_id, visit.filter)
+                src_bucket.copy_blob(src_blob, dest_bucket, new_name=fname)
+            process_group(publisher, visit_infos, upload_from_pool)
+            _log.info("Slewing to next group")
+            time.sleep(SLEW_INTERVAL)
+    else:
+        _log.info(f"No raw files found for {instrument}, generating dummy files instead.")
+        for i in range(n_groups):
+            group = last_group + i + 1
+            visit_infos = make_random_visits(instrument, group)
+
+            # TODO: may be cleaner to use a functor object than to depend on
+            # closures for the bucket and data.
+            def upload_dummy(visit, snap_id):
```

Comment:
Maybe `upload_faked_files` instead of dummy? Or does that not even upload files at all, so just `upload_faked_visits_no_files` (which is too long, but...)

Reply:
This does, in fact, upload a file -- most of the contents of `-main` were uploaded using this code (pre-refactor, of course).